firedrake/preconditioners/patch.py

```diff
@@ -602,8 +603,12 @@ def Ffacet_op(pc, point, vec, mat, facetIS, facet_dofmap, facet_dofmapWithAll):
                                            dtype=IntType)
                     nfacet = len(facets)
                     dofs = facet_dofmap.ctypes.data
-                    Fint_facet_kernel.funptr(0, nfacet, facets.ctypes.data, mat.handle,
-                                             dofs, dofs, *facet_Fop_args)
+                    if facet_Fop_state_data_slot is not None:
+                        assert facet_dofmapWithAll is not None
+                        facet_Fop_data_args[facet_Fop_state_data_slot] = vec.array_r.ctypes.data
+                        facet_Fop_map_args[facet_Fop_state_map_slot] = facet_dofmapWithAll.ctypes.data
+                    Fint_facet_kernel.funptr(0, nfacet, facets.ctypes.data, mat.handle, *facet_Fop_data_args,
+                                             dofs, *facet_Fop_map_args)
```

Comment:
I need to think about how to refactor `assemble` to give us a bunch of this "for free", but have not done so yet.